src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrE2ETestsBase.cs

```diff
@@ -0,0 +1,274 @@
+
+        if (directory.Name == ".git")
+        {
+            return files;
+        }
+
+        files.AddRange(directory.GetFiles().Select(f => new NativePath(f.FullName)));
+
+        foreach (var subDir in directory.GetDirectories())
+        {
+            files.AddRange(GetAllFilesInDirectory(subDir));
+        }
+
+        return files;
+    }
+
+    protected string GenerateSourceMappings(
+        ICollection<SourceMapping> mappings,
+        string patchesPath = "",
+        ICollection<AdditionalMapping>? additionalMappings = null,
+        ICollection<string>? exclude = null)
+    {
+        var additionalMappingsString = string.Empty;
+        var mappingsString = string.Empty;
+
+        if(additionalMappings != null)
+        {
+            additionalMappingsString = string.Join(
+                "," + Environment.NewLine, 
+                additionalMappings.Select(m => 
+                string.Format(Constants.AdditionalMappingTemplate, new[] { m.Source, m.Destination })));
+        }
+
+        mappingsString = string.Join("," + Environment.NewLine, mappings.Select(m => GenerateMappingString(m)));
+        return string.Format(Constants.SourceMappingsTemplate, new[] { patchesPath, additionalMappingsString, mappingsString });
+    }
+
+    private string GenerateMappingString(SourceMapping mapping)
+    {
+        var excluded = string.Empty;
+
+        if (mapping.Exclude != null)
+        {
+            excluded = string.Join(", ", mapping.Exclude.Select(e => $"\"{e}\""));
+        }
+
+        return string.Format(Constants.MappingTemplate, new[] { mapping.Name, mapping.DefaultRemote, excluded });
+    }
+
+    protected async Task<string> CopyRepoAndCreateVersionDetails(
+        LocalPath currentTestDir,
+        string repoName,
+        IDictionary<string, List<Dependency>>? dependencies = null)
+    {
+        var repoPath = currentTestDir / repoName;
+        CopyDirectory(VmrE2ETestsOneTimeSetUp.testsDirectory / repoName, repoPath);
+        
+        var dependenciesString = new StringBuilder();
+        if (dependencies != null && dependencies.ContainsKey(repoName))
+        {
+            var repoDependencies = dependencies[repoName];
+            foreach (var dep in repoDependencies)
+            {
+                string sha = await CopyRepoAndCreateVersionDetails(currentTestDir, dep.Name, dependencies);
+                dependenciesString.AppendLine(
+                    string.Format(
+                        Constants.DependencyTemplate,
+                        new[] { dep.Name, EscapePath(dep.Uri), sha }));
+            }
+        }
+
+        var versionDetails = string.Format(Constants.VersionDetailsTemplate, dependenciesString);
+        File.WriteAllText(repoPath / "eng" / _versionDetailsName, versionDetails);
+        await CommitAll(repoPath, "update version details");
+        return await GetRepoLastCommit(repoPath);
+    }
+
+    public static string EscapePath(string path)
```

Comment:
I'd turn this into an extension method and then you can do:
```
var path = "....";
path = path.Escape();
```

Also I see this escaping all over the place so you please go through those and use this method.

I am also curious why we need to keep escaping these?